// @spectrum-web-components/accordion/src/Accordion.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from "tslib";
import { html, property, queryAssignedNodes, } from '@spectrum-web-components/base';
import styles from './accordion.css.js';
import { Focusable, getActiveElement } from '@spectrum-web-components/shared';
/**
 * @element sp-accordion
 */
export class Accordion extends Focusable {
    constructor() {
        super(...arguments);
        /**
         * Allows multiple accordion items to be opened at the same time
         */
        this.allowMultiple = false;
    }
    static get styles() {
        return [styles];
    }
    get items() {
        return [...(this.defaultNodes || [])].filter((node) => typeof node.tagName !== 'undefined');
    }
    focus() {
        if (this.focusElement === this) {
            return;
        }
        super.focus();
    }
    /**
     * @private
     */
    get focusElement() {
        const items = this.items;
        if (items && !items.length) {
            return this;
        }
        let index = 0;
        while (index < items.length && items[index] && items[index].disabled) {
            index += 1;
        }
        if (items[index]) {
            return items[index];
        }
        /* c8 ignore next */
        return this;
    }
    startListeningToKeyboard() {
        const items = this.items;
        /* c8 ignore next 3 */
        if (items && !items.length) {
            return;
        }
        this.addEventListener('keydown', this.handleKeydown);
    }
    stopListeningToKeyboard() {
        this.removeEventListener('keydown', this.handleKeydown);
    }
    handleKeydown(event) {
        const { code } = event;
        /* c8 ignore next 3 */
        if (code !== 'ArrowDown' && code !== 'ArrowUp') {
            return;
        }
        event.preventDefault();
        const direction = code === 'ArrowDown' ? 1 : -1;
        this.focusItemByOffset(direction);
    }
    focusItemByOffset(direction) {
        const items = this.items;
        const focused = items.indexOf(getActiveElement(this));
        let next = focused;
        let availableItems = items.length;
        // cycle through the available items in the directions of the offset to find the next non-disabled item
        while ((items[next].disabled || next === focused) && availableItems) {
            availableItems -= 1;
            next = (items.length + next + direction) % items.length;
        }
        // if there are no non-disabled items, skip the work to focus a child
        if (items[next].disabled || next === focused) {
            return;
        }
        items[next].focus();
    }
    onToggle(event) {
        const target = event.target;
        const items = [...this.items];
        /* c8 ignore next 3 */
        if (items && !items.length) {
            return;
        }
        if (!this.allowMultiple && !event.defaultPrevented) {
            items.forEach((item) => {
                if (item.open && item !== target) {
                    item.open = false;
                }
            });
        }
    }
    render() {
        return html `
            <slot></slot>
        `;
    }
    firstUpdated(changed) {
        super.firstUpdated(changed);
        this.addEventListener('focusin', this.startListeningToKeyboard);
        this.addEventListener('focusout', this.stopListeningToKeyboard);
        this.addEventListener('sp-accordion-item-toggle', this.onToggle);
    }
}
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'allow-multiple' })
], Accordion.prototype, "allowMultiple", void 0);
__decorate([
    queryAssignedNodes()
], Accordion.prototype, "defaultNodes", void 0);
